service/listenerpipe.go

```diff
@@ -0,0 +1,53 @@
+package service
+
+import (
+	"errors"
+	"net"
+	"sync"
+)
+
+// ListenerPipe returns a full-duplex in-memory connection, like net.Pipe.
+// Unlike net.Pipe one end of the connection is returned as an object
+// satisfying the net.Listener interface.
+// The first call to the Accept method of this object will return a net.Conn
+// connected to the other net.Conn returned by ListenerPipe.
+// Any subsequent calls to Accept will block until the listener is closed.
+func ListenerPipe() (net.Listener, net.Conn) {
+	conn0, conn1 := net.Pipe()
+	return &preconnectedListener{conn: conn0, closech: make(chan struct{})}, conn1
+}
+
+// preconnectedListener satisfies the net.Listener interface by accepting a
+// single pre-established connection.
+// The first call to Accept will return the conn field, any subsequent call
+// will block until the listener is closed.
+type preconnectedListener struct {
+	done    bool
+	conn    net.Conn
+	closech chan struct{}
+	closeMu sync.Mutex
+}
+
+func (l *preconnectedListener) Accept() (net.Conn, error) {
+	if !l.done {
+		l.done = true
+		return l.conn, nil
+	}
+	<-l.closech
+	return nil, errors.New("closed")
```

Comment:
Let's return a more descriptive error here. Something like: `accept failed: listener closed`.

Reply:
Alternatively we could create and return an `OpError` which is used in the net package: https://golang.org/pkg/net/#OpError. It may not be as useful since this is all local anyways, but would be consistent.

Reply:
I think we would need to fake too many fields, there's no network type and no address.